### PR TITLE
Remove support of C++11 from oneDPL

### PR DESCRIPTION
**Pull request by SergeyKopienko**

### CODE REVIEW

**Comment by SergeyKopienko** [comment]
If we will ready to apply this patch, required to remove C++11 from CI and nightly tests.

**Comment by SergeyKopienko** [comment]
CI tasks
- https://github.com/oneapi-src/oneDPL/actions/runs/3088797079/jobs/4995696020
- https://github.com/oneapi-src/oneDPL/actions/runs/3088797079/jobs/4995696433
was failed because they use C++11 standard.

File: include/oneapi/dpl/internal/common_config.h [hunk 1]
```diff
@@ -16,7 +16,7 @@
 #ifndef _ONEDPL_COMMON_CONFIG_H
 #define _ONEDPL_COMMON_CONFIG_H
 // Workarounds for libstdc++9, libstdc++10 when new TBB version is found in the environment
-#if __cplusplus >= 201703L
+#if __cplusplus >= 201703L                                                  // KSATODO: leve as is: https://www.intel.com/content/www/us/en/developer/articles/release-notes/intel-oneapi-threading-building-blocks-release-notes.html
```
Comment (akukanov) [inline include/oneapi/dpl/internal/common_config.h:19]: Same again, the check guards `__has_include` and is supposed to always succeed with C++17.

File: include/oneapi/dpl/internal/serial_algorithm_impl.h [hunk 2]
```diff
@@ -18,8 +18,8 @@
 #define _ONEDPL_SERIAL_NUMERIC_IMPL_H
 
 #if (__cplusplus >= 201703L)
-#    include <algorithm>
-#    include <iterator>
+#    include <algorithm>            // KSATODO: we may include this file without __cplusplus check
+#    include <iterator>             // KSATODO: we may include this file without __cplusplus check
```
Comment (akukanov) [inline include/oneapi/dpl/internal/serial_algorithm_impl.h:22]: The idea of this code block (lines 20-46) was to add a serial version of `for_each_n` when it is needed (C++17 and above) but not available (`_ONEDPL_HAS_NUMERIC_SERIAL_IMPL`). With C++17 being the minimal supported version, the algorithms is always needed, but still not always available.

File: include/oneapi/dpl/internal/serial_numeric_impl.h [hunk 3]
```diff
@@ -18,9 +18,9 @@
 #define _ONEDPL_SERIAL_NUMERIC_IMPL_H
 
 #if (__cplusplus >= 201703L)
-#    include <functional>
-#    include <iterator>
-#    include <numeric>
+#    include <functional>       // KSATODO: we may include this file without __cplusplus check https://en.cppreference.com/w/cpp/utility/functional/function - since C++11
+#    include <iterator>         // KSATODO: we may include this file without __cplusplus check https://en.cppreference.com/w/cpp/header/iterator
+#    include <numeric>          // KSATODO: we may include this file without __cplusplus check https://en.cppreference.com/w/cpp/header/numeric
```
Comment (akukanov) [inline include/oneapi/dpl/internal/serial_numeric_impl.h:23]: Same thing as in `serial_algorithm_impl.h`

File: include/oneapi/dpl/iterator [hunk 4]
```diff
@@ -34,19 +34,17 @@ namespace dpl
...
 using ::std::rbegin;
 using ::std::rend;
-#endif // __cplusplus > 201103L
-using ::std::distance;
 #if __cplusplus > 201402L
-using ::std::empty;
-using ::std::size;
+using ::std::empty;         // KSATODO: leve as is: https://en.cppreference.com/w/cpp/iterator/empty - since С++17
+using ::std::size;          // KSATODO: leve as is: https://en.cppreference.com/w/cpp/iterator/size - since С++17
 #endif // __cplusplus > 201402L
```
Comment (akukanov) [inline include/oneapi/dpl/iterator:48]: Why keeping the "greater than C++14" condition check?

File: include/oneapi/dpl/optional [hunk 5]
```diff
@@ -24,10 +24,10 @@ namespace oneapi
...
 #if __cplusplus > 201402L
-using ::std::make_optional;
-using ::std::nullopt;
-using ::std::nullopt_t;
-using ::std::optional;
+using ::std::make_optional;     // KSATODO: leave as is: https://en.cppreference.com/w/cpp/utility/optional/make_optional - since C++17
+using ::std::nullopt;           // KSATODO: leave as is: https://en.cppreference.com/w/cpp/utility/optional/nullopt - since C++17
+using ::std::nullopt_t;         // KSATODO: leave as is: https://en.cppreference.com/w/cpp/types/nullptr_t - since C++17
+using ::std::optional;          // KSATODO: leave as is: https://en.cppreference.com/w/cpp/utility/optional - since C++17
 #endif // __cplusplus > 201402L
```
Comment (akukanov) [inline include/oneapi/dpl/optional:31]: Why keeping the "greater than C++14" condition check?

**Review by akukanov — COMMENTED** [review]
First portion of comments, to be continued.

File: include/oneapi/dpl/execution [hunk 6]
```diff
@@ -20,7 +20,7 @@
 // ATTENTION!!! Include the header from the C++ standard library before the oneDPL config header to see whether the _PSTL_VERSION macro is defined
 #if __cplusplus >= 201703L
 #    if __has_include(<execution>)
-#        include <execution>
+#        include <execution>                    // KSATODO: leve as is: https://en.cppreference.com/w/cpp/header/execution - Standard library header <execution> (C++17)
 #    endif // __has_include(<execution>)
 #endif     // __cplusplus >= 201703L
```
[the remark this answers is not in the document]
Comment (SergeyKopienko) [reply include/oneapi/dpl/execution:25]: Fixed: removed support of C++14

File: include/oneapi/dpl/functional [hunk 7]
```diff
@@ -48,19 +48,16 @@ using ::std::plus;
 using ::std::ref;
 using ::std::reference_wrapper;
 #if __cplusplus < 201703L
-using ::std::binary_function;
-using ::std::unary_function;
+using ::std::binary_function;   // KSATODO: leve as is: https://en.cppreference.com/w/cpp/utility/functional/binary_function - deprecated in C++11, removed in C++17
+using ::std::unary_function;    // https://en.cppreference.com/w/cpp/utility/functional/unary_function - deprecated in C++11, removed in C++17
 #endif // __cplusplus < 201703L
```
[the remark this answers is not in the document]
Comment (SergeyKopienko) [reply include/oneapi/dpl/functional:53]: Fixed: removed support of C++14

File: include/oneapi/dpl/functional [hunk 8]
```diff
@@ -48,19 +48,16 @@ using ::std::plus;
...
-using ::std::not2;
+using ::std::binary_negate;     // KSATODO: leve as is: https://en.cppreference.com/w/cpp/utility/functional/binary_negate - deprecated in C++17
+using ::std::not1;              // KSATODO: leve as is: https://en.cppreference.com/w/cpp/utility/functional/not1 - since C++14, deprecated in C++17, removed in C++20
+using ::std::not2;              // KSATODO: leve as is: https://en.cppreference.com/w/cpp/utility/functional/not2 - deprecated in C++17, removed in C++20
 using ::std::unary_negate;
 #endif // __cplusplus < 202002L
-
-#if __cplusplus > 201103L
 using ::std::bit_not;
-#endif // __cplusplus > 201103L
```
[the remark this answers is not in the document]
Comment (SergeyKopienko) [reply include/oneapi/dpl/functional:63]: Fixed: removed support of C++14 too

File: include/oneapi/dpl/internal/serial_numeric_impl.h (hunk 3 above)
Comment (SergeyKopienko) [reply include/oneapi/dpl/internal/serial_numeric_impl.h:23]: Fixed: removed support of C++14 too

File: include/oneapi/dpl/iterator (hunk 4 above)
Comment (SergeyKopienko) [reply include/oneapi/dpl/iterator:48]: Fixed: removed support of C++14 too

File: include/oneapi/dpl/optional (hunk 5 above)
Comment (SergeyKopienko) [reply include/oneapi/dpl/optional:31]: Fixed: removed support of C++14 too

File: include/oneapi/dpl/functional [hunk 9]
```diff
@@ -48,19 +48,16 @@ using ::std::plus;
...
 #endif // __cplusplus < 201703L
 #if __cplusplus < 202002L
-using ::std::binary_negate;
-using ::std::not1;
-using ::std::not2;
+using ::std::binary_negate;     // KSATODO: leve as is: https://en.cppreference.com/w/cpp/utility/functional/binary_negate - deprecated in C++17
+using ::std::not1;              // KSATODO: leve as is: https://en.cppreference.com/w/cpp/utility/functional/not1 - since C++14, deprecated in C++17, removed in C++20
+using ::std::not2;              // KSATODO: leve as is: https://en.cppreference.com/w/cpp/utility/functional/not2 - deprecated in C++17, removed in C++20
 using ::std::unary_negate;
 #endif // __cplusplus < 202002L
```
[the remark this answers is not in the document]
Comment (SergeyKopienko) [reply include/oneapi/dpl/functional:59]: Add comment in this place and in other places like this:
**// Deprecated in C++17, removed in C++20**

**Comment by SergeyKopienko** [comment]
@akukanov Continue to remove C++17 in PR https://github.com/oneapi-src/oneDPL/pull/642
This PR not actual anymore.

File: include/oneapi/dpl/internal/common_config.h (hunk 1 above)
Comment (SergeyKopienko) [reply include/oneapi/dpl/internal/common_config.h:19]: Fixed.

File: include/oneapi/dpl/internal/serial_algorithm_impl.h (hunk 2 above)
Comment (SergeyKopienko) [reply include/oneapi/dpl/internal/serial_algorithm_impl.h:22]: OK.

File: include/oneapi/dpl/functional (hunk 7 above)
[the remark this answers is not in the document]
Comment (SergeyKopienko) [reply include/oneapi/dpl/functional:53]: Removed.

**Pull request closed without merging**